### PR TITLE
FEA: update the parameters of data filter

### PR DESCRIPTION
**Pull request by ChangxinTian**

### CODE REVIEW

File: recbole/data/dataset/dataset.py [hunk 1]
```diff
@@ -667,80 +657,115 @@ def _filter_by_inter_num(self):
...
+        if intervals_str is None:
+            return None
 
-    def _drop_by_value(self, val, cmp):
-        """Drop illegal rows by value.
+        endpoints = []
+        for endpoint_pair_str in str(intervals_str).split(';'):
+            left_bracket, right_bracket = endpoint_pair_str[0], endpoint_pair_str[-1]
+            endpoint_pair = endpoint_pair_str.strip('[]()').split(',')
+            if not (len(endpoint_pair)==2 and left_bracket in ['(','['] and right_bracket in [')',']']):
```
Comment (hyp1231) [inline recbole/data/dataset/dataset.py:707]:
```suggestion
            if not (len(endpoint_pair) == 2 and left_bracket in ['(', '['] and right_bracket in [')', ']']):
```

File: recbole/data/dataset/dataset.py [hunk 2]
```diff
@@ -667,80 +657,115 @@ def _filter_by_inter_num(self):
...
+                self.logger.warning( f'{endpoint_pair_str} is an illegal interval!')
+                continue
 
-        Args:
-            val (dict): value that compared to.
-            cmp (Callable): return False if a row need to be dropped
+            def str2npnum(num_str):
+                if num_str in ["inf","-inf"]:
```
Comment (hyp1231) [inline recbole/data/dataset/dataset.py:712]:
```suggestion
                if num_str in ["inf", "-inf"]:
```

File: recbole/data/dataset/dataset.py [hunk 3]
```diff
@@ -667,80 +657,115 @@ def _filter_by_inter_num(self):
...
+        Note:
+            return true when the intervals is None.
+        """
+        if intervals is None:
+            return True
+
+        result = None
+        for left_bracket, left_point, right_point, right_bracket in intervals:
+            temp_result = num>=left_point if left_bracket=='[' else num>left_point
+            temp_result &= num<=right_point if right_bracket==']' else num<right_point
```
Comment (hyp1231) [inline recbole/data/dataset/dataset.py:738]:
```suggestion
            temp_result = num >= left_point if left_bracket == '[' else num > left_point
            temp_result &= num <= right_point if right_bracket == ']' else num < right_point
```

**Review by hyp1231 — APPROVED** [review]
Others,

* `_analyse_intervals_str` or `_parse_intervals_str`, which one is better? (actually I'm not sure.)

**Comment by ChangxinTian** [comment]
> Others,
> 
> * `_analyse_intervals_str` or `_parse_intervals_str`, which one is better? (actually I'm not sure.

`_parse_intervals_str` is better.
I will carefully check the code format in the future, thank you!

File: docs/source/user_guide/data/data_args.rst [hunk 4]
```diff
@@ -72,10 +72,7 @@ Remove duplicated user-item interactions
 Filter by value
 ''''''''''''''''''
 
-- ``lowest_val (dict)`` : Has the format ``{k (str): v (float)}, ...``. The rows whose ``feat[k] < v`` will be filtered. Defaults to ``None``.
-- ``highest_val (dict)`` : Has the format ``{k (str): v (float)}, ...``. The rows whose ``feat[k] > v`` will be filtered. Defaults to ``None``.
-- ``equal_val (dict)`` : Has the format ``{k (str): v (float)}, ...``. The rows whose ``feat[k] != v`` will be filtered. Defaults to ``None``.
-- ``not_equal_val (dict)`` : Has the format ``{k (str): v (float)}, ...``. The rows whose ``feat[k] == v`` will be filtered. Defaults to ``None``.
+- ``val_interval (dict)``: Has the format ``{k (str): interval (str), ...}``, where ``interval `` can be set as  ``[A,B]`` / ``[A,B)`` / ``(A,B)`` / ``(A,B]``. The rows whose ``feat[k]`` is in the interval ``interval`` will be retained. If you want to specify more than one interval, separate them with semicolon(s). For instance, ``{k: [A,B);(C,D]}`` can be adopted and rows whose ``feat[k]`` is in any specified interval will be retained. Defaults to ``None``, which means all rows will be retained.
```
Comment (chenyushuo) [inline docs/source/user_guide/data/data_args.rst:75]: ``{k: [A,B);(C,D]}`` -> ``{k: "[A,B);(C,D]"}``

File: docs/source/user_guide/data/data_args.rst [hunk 5]
```diff
@@ -85,10 +82,8 @@ Remove interation by user or item
...
 ''''''''''''''''''''''''''''''''''''
 
-- ``max_user_inter_num (int)`` : Users whose number of interactions is more than ``max_user_inter_num`` will be filtered. Defaults to ``None``.
-- ``min_user_inter_num (int)`` : Users whose number of interactions is less than ``min_user_inter_num`` will be filtered. Defaults to ``0``.
-- ``max_item_inter_num (int)`` : Items whose number of interactions is more than ``max_item_inter_num`` will be filtered. Defaults to ``None``.
-- ``min_item_inter_num (int)`` : Items whose number of interactions is less than ``min_item_inter_num`` will be filtered. Defaults to ``0``.
+- ``user_inter_num_interval (str)`` : Has the interval format, such as ``[A,B]`` / ``[A,B)`` / ``(A,B)`` / ``(A,B]``,  where ``A`` and ``B`` are the endpoints of the interval and ``A<B``. Users whose number of interactions is in the interval will be retained. Defaults to ``[0,inf)``.
+- ``item_inter_num_interval (str)`` : Has the interval format, such as ``[A,B]`` / ``[A,B)`` / ``(A,B)`` / ``(A,B]``,  where ``A`` and ``B`` are the endpoints of the interval and ``A<B``. Items whose number of interactions is in the interval will be retained. Defaults to ``[0,inf)``.
```
Comment (chenyushuo) [inline docs/source/user_guide/data/data_args.rst:86]: ``A<B`` -> ``A <= B``

File: recbole/data/dataset/dataset.py [hunk 6]
```diff
@@ -667,80 +657,115 @@ def _filter_by_inter_num(self):
...
-            feat.reset_index(drop=True, inplace=True)
+        if intervals_str is None:
+            return None
 
-    def _drop_by_value(self, val, cmp):
-        """Drop illegal rows by value.
+        endpoints = []
+        for endpoint_pair_str in str(intervals_str).split(';'):
+            left_bracket, right_bracket = endpoint_pair_str[0], endpoint_pair_str[-1]
+            endpoint_pair = endpoint_pair_str.strip('[]()').split(',')
```
Comment (chenyushuo) [inline recbole/data/dataset/dataset.py:706]: 这里建议改成：
```python3
            endpoint_pair_str = endpoint_pair_str.strip()
            left_bracket, right_bracket = endpoint_pair_str[0], endpoint_pair_str[-1]
            endpoint_pair = endpoint_pair_str[1:-1].split(',')
```

File: recbole/data/dataset/dataset.py [hunk 7]
```diff
@@ -667,80 +657,115 @@ def _filter_by_inter_num(self):
...
+            if not (len(endpoint_pair) == 2 and left_bracket in ['(', '['] and right_bracket in [')', ']']):
+                self.logger.warning( f'{endpoint_pair_str} is an illegal interval!')
+                continue
 
-        Args:
-            val (dict): value that compared to.
-            cmp (Callable): return False if a row need to be dropped
+            def str2npnum(num_str):
+                if num_str in ["inf", "-inf"]:
```
Comment (chenyushuo) [inline recbole/data/dataset/dataset.py:712]: 建议``num_str``改为``num_str.lower()``

File: recbole/data/dataset/dataset.py [hunk 8]
```diff
@@ -667,80 +657,115 @@ def _filter_by_inter_num(self):
...
+    def _within_intervals(self, num, intervals):
+        """ return Ture if the num is in the intervals.
+
+        Note:
+            return true when the intervals is None.
+        """
+        if intervals is None:
+            return True
+
+        result = None
```
Comment (chenyushuo) [inline recbole/data/dataset/dataset.py:735]: ``result``的初值改为``True``，顺便修改下面的``result``更新表达式

**Review by chenyushuo — CHANGES_REQUESTED** [review]

File: recbole/data/dataset/dataset.py [hunk 9]
```diff
@@ -667,80 +657,113 @@ def _filter_by_inter_num(self):
...
+                if num_str.lower() in ["inf", "-inf"]:
+                    return np.inf if num_str == "inf" else -np.inf
+                else:
+                    try:
+                        return float(num_str)
+                    except ValueError:
+                        raise ValueError(f'Str {num_str} in interval can not be converted to numeric.')
+            left_point, right_point = str2npnum(endpoint_pair[0]), str2npnum(endpoint_pair[1])
+            if left_point > right_point:
+                self.logger.warning( f'{endpoint_pair_str} is an illegal interval!')
```
Comment (chenyushuo) [inline recbole/data/dataset/dataset.py:722]: 这里多了一个空格

File: recbole/data/dataset/dataset.py [hunk 10]
```diff
@@ -667,80 +657,113 @@ def _filter_by_inter_num(self):
...
+            if not (len(endpoint_pair) == 2 and left_bracket in ['(', '['] and right_bracket in [')', ']']):
+                self.logger.warning( f'{endpoint_pair_str} is an illegal interval!')
+                continue
 
-        Args:
-            val (dict): value that compared to.
-            cmp (Callable): return False if a row need to be dropped
+            def str2npnum(num_str):
+                if num_str.lower() in ["inf", "-inf"]:
+                    return np.inf if num_str == "inf" else -np.inf
```
Comment (chenyushuo) [inline recbole/data/dataset/dataset.py:714]: 这里也加一下 .lower()

**Review by chenyushuo — CHANGES_REQUESTED** [review]

**Review by chenyushuo — APPROVED** [review]

**Pull request merged**